[수정][리팩토링][SCH 수정4차]

smilesw6448 opened the pull request:
1. Filter Class 생성하여 SearchCommand.java 역할을 줄임
2. 옵션이 있는 항목별로 Filter Class 생성
3. 관리차원에서 as-was as-is 차이 확인 부탁드립니다.
4.이전 ReqObj에 대한 논의는 아래 계속 진행 필요

@WhiteLeeDev WhiteLeeDev 3 hours ago
RequestObj는 다른 커맨드 객체에서 쓸 수 있을듯 한데요~ common package(만들어야 합니다)에 들어가면 좋지 않을까요?ㅎㅎ
그리고 naming 관련하여 UserRequest가 있는 상황에서 RequestObj라는 것이 하나 더 있어서 나중에 좀 헛갈릴 수 있을듯 합니다 RequestObj의 역할은 user의 input에서 command가 사용할 컬럼, value,option만 뽑아 한번 더 정제한 것이니....... 뭐가 좋을까요?ㅎㅎ

Collaborator
Author
@smilesw6448 smilesw6448 1 hour ago •
맞습니다 가독성을 위해 아래 변수들을 선언해놓고 사용해야 좋은데,
String column = request.getArguments().get(0);
String value = request.getArguments().get(1);
String option = request.getOptions().get(0);
두 질문을 한꺼번에 처리할 수 있는 방법은 ReqestObj Class 삭제 -> 위 변수들을 private으로 선언해서 사용하면 어떨까요?

Collaborator
Author
@smilesw6448 smilesw6448 1 hour ago
제 생각을 덧붙이자면, 일단 obj를 만든이유는 아래와 같습니다.

Argument 인자로 3개를 넘기는것 보다 Object 1개를 넘기는게 더 깔끔해 보이기 때문
가독성을 위해 request를 한 번 정제는 해야 했음

smilesw6448 committed:
[수정][리팩토링][SCH 수정4차] Filter Class 생성하여
SearchCommand.java 역할을 줄임

이전 ReqObj에 대한 논의는 아래 계속 진행 필요

@WhiteLeeDev WhiteLeeDev 3 hours ago
RequestObj는 다른 커맨드 객체에서 쓸 수 있을듯 한데요~ common package(만들어야 합니다)에 들어가면 좋지 않을까요?ㅎㅎ
그리고 naming 관련하여 UserRequest가 있는 상황에서 RequestObj라는 것이 하나 더 있어서 나중에 좀 헛갈릴 수 있을듯 합니다 RequestObj의 역할은 user의 input에서 command가 사용할 컬럼, value,option만 뽑아 한번 더 정제한 것이니....... 뭐가 좋을까요?ㅎㅎ

Collaborator
Author
@smilesw6448 smilesw6448 1 hour ago •
맞습니다 가독성을 위해 아래 변수들을 선언해놓고 사용해야 좋은데,
String column = request.getArguments().get(0);
String value = request.getArguments().get(1);
String option = request.getOptions().get(0);
두 질문을 한꺼번에 처리할 수 있는 방법은 ReqestObj Class 삭제 -> 위 변수들을 private으로 선언해서 사용하면 어떨까요?

Collaborator
Author
@smilesw6448 smilesw6448 1 hour ago
제 생각을 덧붙이자면, 일단 obj를 만든이유는 아래와 같습니다.

Argument 인자로 3개를 넘기는것 보다 Object 1개를 넘기는게 더 깔끔해 보이기 때문
가독성을 위해 request를 한 번 정제는 해야 했음

happy-dev-travel commented on src/main/java/command/sch/EmployeeBirthdayFilter.java:
```diff
@@ -0,0 +1,80 @@
+        }
+    }
+
+    public List<Employee> searchByBirthdayYear(List<Employee> searchedEmployees) {
+
+        List<Employee> resultList = new ArrayList<>();
+        for (Iterator<Employee> itr = searchedEmployees.iterator(); itr.hasNext(); ) {
+            Employee itrEmployee = itr.next();
+            String itrBirthday = itrEmployee.getBirthday();
+            String itrBirthdayYear = itrBirthday.substring(0, 4);
+
+            if (this.value.equals(itrBirthdayYear)) {
+                resultList.add(itrEmployee);
+            }
+        }
+        return resultList;
```
searchByBirthdayYear,searchByBirthdayMonth, searchByBirthdayDay  3개의 함수에서 중복 코드가 발생하고 있습니다 코드가 다른 부분은 birthday에서 필요한 부분을 뽑아내는 substring의 인자뿐인데요. 유지보수성을 위해선 중복코드가 최소호되 하니, 3개의 함수를 하나로 만들고 substring하는 동작만 주입해 주면 좋을 것 같습니다👍👍

smilesw6448 replied:
네네 이부분은 안그래도 중복이 있어서 제거하려고 했는데
리팩토링 해보겠습니다!

happy-dev-travel commented on src/main/java/command/sch/EmployeeNameFilter.java:
```diff
@@ -0,0 +1,67 @@
+        for (Iterator<Employee> itr = searchedEmployees.iterator(); itr.hasNext(); ) {
+            Employee itrEmployee = itr.next();
+            String itrFullName = itrEmployee.getName();
+            String itrFirstName = itrFullName.substring(0, itrFullName.indexOf(" "));
```
여기서도 같은 중복이 발생하고 있는데요, 옵션별로 데이터를 조작하는 부분만 함수에 넘겨주면 전반적인 중복코드가 모두 제거되어 유지보수성이 향상될 것 같습니다 :)

junheui27 replied:
저도 같은 생각입니다 초기화 시에 클래스 별로 중복 발생하는 거를 한번만 하게 해주면 깔끔해질 것 같아요.

happy-dev-travel commented on src/main/java/command/sch/EmployeeBirthdayFilter.java:
```diff
@@ -0,0 +1,80 @@
+        this.option = option;
+    }
+
+    public List<Employee> process(List<Employee> searchedEmployees) {
```
여기 @Override 키워드가 있어야하지 않나요?? 없어도 정상동작 하나요?

smilesw6448 replied:
네 없어도 동작은하나, 가독성을 위해 넣어주었어야 했는데 추가하겠습니다!

happy-dev-travel commented on src/main/java/command/sch/EmployeeBirthdayFilter.java:
```diff
@@ -0,0 +1,80 @@
+    private String value;
+    private String option;
+    EmployeeBirthdayFilter(String column, String value, String option){
+        super(column, value, option);
```
동일한 필드 초기화가 부모클래스인 SearchFilter와 sub클래스들에서 두번 발생하고 있는데요, SearchFilter의 필드들을 protected로 사용하거나, SearchFilter의 기본동작도 하위클래스로 만들고 Filter들이 SearchFilterI 인터페이스를 바로 구현하도록 하면 좋을 듯 합니다 :)

happy-dev-travel reviewed (approved):
기조에 Extract Method를 적용하기에 용이하게 개발을 해주셔서 깔끔하게 SearchCommand의 역할이 축소되고 개별 Filter 기능별로 SRP가 달성된 것 같습니다👍👍 몇가지 코멘트 드린 부분들이 있는데 나름 코드 변경이 꽤 될 것 같아 다음 PR에 반영을 고려해 주시면 감사하겠습니다

jw451 commented on src/main/java/command/sch/EmployeeBirthdayFilter.java:
```diff
@@ -0,0 +1,80 @@
+        return resultList;
+    }
+
+    public List<Employee> searchByBirthdayMonth(List<Employee> searchedEmployees) {
```
iterator 사용도 좋군요. 아래와 같은방식도 가독성이나쁘지않을것같습니다. 
for(Employee itrEmployee : searchedEmployees) {
   itrBirthday=itrEmployee.getBirthday();
  //이하생략
 }

smilesw6448 replied:
오 너무 좋은생각입니다. 제가 왜 저 구문을 생각하고 있지 않았을까요... Iterator가 익숙하다보니..
서지원님 의견이 훨씬 더 깔끔해 보입니다 ㅎㅎ 이건 바로 수정해야겠어요

jw451 reviewed (approved):
SCH가 복잡하고 할것도 많은데 고생많으십니다.
리뷰했습니다.

junheui27 commented on src/main/java/command/sch/SearchCommand.java:
```diff
@@ -36,223 +34,39 @@ public List<Employee> run(UserRequest request, EmployeeDB db) {
 
     private RequestObj makeReqObj(UserRequest request) {
-        String key = request.getArguments().get(0);
+        String column = request.getArguments().get(0);
```
옵션의 경우 제가 converter 구현한거에서는 0번째 인자가 p로 들어가는데 이부분은 같이 얘기해서 수정해야 할 것 같습니다.

smilesw6448 committed:
[수정][리팩토링][SCH 수정5차]
1. 중복로직 제거
=> SearchFilterI -> findValueByOption, searchByValue 함수 생성

2. @OVerride 추가

3. iterator -> for 문으로 깔끔하게 변경

4. option 부분 가져오는 인자 1로 변경 (0->1)
